mongocrypt/src/native.rs

```diff
@@ -0,0 +1,47 @@
+use std::borrow::Borrow;
+
+pub(crate) type Destroy<T> = unsafe extern "C" fn(*mut T);
+
+pub(crate) struct OwnedPtr<T> {
+    inner: *mut T,
+    destroy: Destroy<T>,
+}
+
+#[cfg(all(test, feature = "compile_fail"))]
+#[test]
+fn use_after_free() {
+    unsafe extern "C" fn destroy(_ptr: *mut ()) {}
+    let owned: OwnedPtr<()> = OwnedPtr::new(std::ptr::null_mut(), destroy);
+    let ptr = owned.borrow();
+    std::mem::drop(owned);
+    unsafe { destroy(*ptr) };
+}
+
+impl<T> OwnedPtr<T> {
+    pub(crate) fn new(inner: *mut T, destroy: Destroy<T>) -> Self {
+        Self { inner, destroy }
+    }
+
+    pub(crate) fn borrow_const(&self) -> &*const T {
+        let borrow: &*mut T = self.borrow();
+        let borrow_ptr = borrow as *const *mut T;
+        let borrow_const_ptr = borrow_ptr as *const *const T;
+        unsafe { &*borrow_const_ptr }
+    }
+}
+
+impl<T> Borrow<*mut T> for OwnedPtr<T> {
```

Comment:
Whether a method on one of the wrapper structs takes `&self` or `&mut self` depends on the specific method; for maximum safety they'd all have to be `&mut` since the C library can do anything when called, but that outright breaks intended usage patterns (repeatedly calling `next_kms_ctx`, using the values of `message` and `endpoint` simultaneously, etc.).  Similar to `BinaryRef`, I've attempted to codify here in the Rust type system what's expressed in the documentation for C.

Reply:
This runs into the same problem as the other - because the C interface, with very few exceptions, takes `*mut` everywhere this would require casts at every usage site.